---
title: "QBIO Final Project: TCGA-HNSC"
---


```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("//wsl.localhost/Ubuntu/root/qbio_490_tritong/analysis_data"))
```


```{r}
# install, only if needed, and load packages (SummarizedExperiment and TCGAbiolinks)
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(version = "3.17")
if (!require("TCGAbiolinks", quietly = TRUE))
  BiocManager::install("TCGAbiolinks")
if (!require("maftools", quietly = TRUE))
  BiocManager::install("maftools")
if (!require("DESeq2", quietly = TRUE))
  BiocManager::install("DESeq2")
if (!require("EnhancedVolcano", quietly = TRUE))
  BiocManager::install("EnhancedVolcano")
if (!require(survival)) {
  install.packages("survival")}
if (!require(survminer)) {
  install.packages("survminer")}
if (!require(mltools)) {
  install.packages("dplyr")}

library(BiocManager)
library(TCGAbiolinks)
library(maftools)
library(ggplot2)
library (SummarizedExperiment)
library(DESeq2)
library(EnhancedVolcano)
library(survminer)
library(survival)
library(dplyr)

```


```{r}
#neccessary downloads and preparation for clinical files

clin_query <- GDCquery(project = "TCGA-HNSC",
                       data.category = "Clinical",
                       data.type = "Clinical Supplement",
                       data.format = 'BCR Biotab')
#GDCdownload(clin_query)
clinical.BCRtab.all <- GDCprepare(clin_query)  
clinic <- clinical.BCRtab.all$clinical_patient_hnsc[-c(1,2),]

colnames(clinic)[ colnames(clinic) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"

write.csv(clinic, "//wsl.localhost/Ubuntu/root/qbio_490_tritong/analysis_data/hnsc_clinical_data.csv", row.names=FALSE)

clinic <- read.csv("//wsl.localhost/Ubuntu/root/qbio_490_tritong/analysis_data/hnsc_clinical_data.csv")


```



```{r}
#neccessary downloads and preparation for rna_se files

rna_query <- GDCquery(project ="TCGA-HNSC",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)

rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)
rna_clinical$Tumor_Sample_Barcode <- rna_clinical$patient

```


```{r}
#neccessary downloads and preparation for MAF files

clinical <- read.csv("//wsl.localhost/Ubuntu/root/qbio_490_tritong/analysis_data/hnsc_clinical_data.csv")
colnames(clinical)[ colnames(clinical) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"
write.csv(clinical, "//wsl.localhost/Ubuntu/root/qbio_490_tritong/analysis_data/hnsc_clinical_data.csv", row.names = FALSE)

maf_query <- GDCquery(
  project = "TCGA-HNSC", 
  data.category = "Simple Nucleotide Variation", 
  access = "open", # we only have access to somatic mutations which are open access
  data.type = "Masked Somatic Mutation", 
  workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking"
)

GDCdownload(maf_query)

maf <- GDCprepare(maf_query) # as long as it runs, ignore any errors

maf_object <- read.maf(maf = maf, 
                       clinicalData = rna_clinical,
                       isTCGA = TRUE)
```


Investigate the different groups of smokers/drinkers
```{r}
#colnames(rna_clinical)

#rna_clinical$cigarettes_per_day

#rna_clinical$alcohol_history

#rna_clinical$pack_years_smoked

mask_alc <- ifelse(rna_clinical$alcohol_history == "Yes", TRUE, FALSE)
mask_noalc <- ifelse(rna_clinical$alcohol_history == "No", TRUE, FALSE)

both <- rna_clinical[mask_alc, ]
both <- both[!is.na(both$cigarettes_per_day), ]

alc_only <- rna_clinical[mask_alc, ]
alc_only <- alc_only[is.na(alc_only$cigarettes_per_day), ]

cig_only <- rna_clinical[mask_noalc, ]
cig_only <- cig_only[!is.na(cig_only$cigarettes_per_day), ]
```

Survival Plot for alc only, cig only, and both
```{r}
### Cleaning and Prepping rna_clinical$definition column for meta and non-metastatic ###
# Remove any patients with no definition information
def_na_mask <- ifelse(is.na(maf_object@clinical.data$definition), F, T)
cleaned_def <- maf_object@clinical.data[def_na_mask,]

# Select only meta and non-metastatic patients
meta_nonmeta_mask <- ifelse(cleaned_def$definition == 'Metastatic' | 
                              cleaned_def$definition == 'Primary solid Tumor', T, F)
cleaned_def <- cleaned_def[meta_nonmeta_mask,]

# Create a survival time column for the survival plot
cleaned_def$survival_time <- ifelse(cleaned_def$days_to_death == 'NA',
                                   cleaned_def$days_to_last_follow_up,
                                   cleaned_def$days_to_death)
# Remove potential -Inf values of survival time
inf_mask <- ifelse(cleaned_def$survival_time == '-Inf', F, T)
cleaned_def <- cleaned_def[inf_mask,]

# Create survival event (T/F) column for survival plots, # T if pt is dead, F if alive
cleaned_def$death_event <- ifelse(cleaned_def$vital_status == 'Alive', F, T)

survival_object <- Surv(time = as.numeric(cleaned_def$survival_time), event = cleaned_def$death_event)

fit_object <- survfit(survival_object ~ definition, data = cleaned_def)

survplot <- ggsurvplot(fit_object, pval = TRUE, ggtheme =
                         theme(plot.margin = unit(c(1,1,1,1), "cm")), legend = 'right')

survplot
```
```

