---
title: "Intro to Transcriptomics"
author: Nicole Black, Wade Boohar, Kayla Xu
date: 07/17/22
updated: 09/26/23
---

***Deliverables***
-Upload this R Notebook to your GitHub and submit the link to your Repo on BB.
-Include any graphs or figures created in this assignment in the folder with your R notebook with descriptive file names.

Since this is an optional partner activity, it is okay if your answers are the same as your partnerâ€™s as long as everyone understands it and could explain it in their own words if asked. Each person must individually push their code to Github. *At the top of your R Notebook, write the name of you and your partner(s) as a comment.*

***Complete the following coding activity and answer any following questions as comments in your R Notebook***

In Transcriptomics_Tutorial.Rmd, you learned how to manipulate the SummarizedExperiment data structure, turn it into more readable dataframes, complete basic analysis using those dataframes, and performed a differential expression analysis based on age. Before saving the rna_counts, rna_clinical, and rna_genes however, we masked out any sample that was not a tumor. In this assignment, you will need to reinstantiate the rna_se object and then perform differential expression analysis based on tumor status.

*Pre-Assignment*
Use knitr function to set your working directory to your analysis_data folder in qbio_490_name.
```{r}
    knitr::opts_knit$set(root.dir = normalizePath("\\\\wsl.localhost/Ubuntu/home/tri/qbio_490_tritong/analysis_data")) 
```

Load all necessary packages and instantiate the rna_se object
```{r}
library(TCGAbiolinks)
library(SummarizedExperiment)
rna_query <- GDCquery(project ="TCGA-BRCA",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
rna_se <- GDCprepare(rna_query)
```


*1*
Perform all data cleaning and label manipulation steps (hint: applying colnames and rownames to make the dataframes more readable) except for masking out the patient with an NA for age and the "Solid Tissue Normal" sample types.
```{r}
#RNA_CLINICAL
rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)
treatments_mask <- ifelse(colnames(rna_clinical) == 'treatments', F, T)
rna_clinical <- rna_clinical[, treatments_mask]
primary_mask <- ifelse(colnames(rna_clinical) == 'primary_site', F, T)
rna_clinical <- rna_clinical[, primary_mask]
diseaseT_mask <- ifelse(colnames(rna_clinical) == 'disease_type', F, T)
rna_clinical <- rna_clinical[, diseaseT_mask]
#RNA_GENES
rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- as.data.frame(rna_genes)
#RNA_COUNTS
rna_counts <- rna_se@assays@data$unstranded
rna_counts <- as.data.frame(rna_counts)
#MAKE AGE CATEGORY
rna_clinical$age_category <- ifelse(rna_clinical$age_at_index <= 58, 'young', 'old')
# Add rows and column names for better readability
rownames(rna_genes) <- rna_genes$gene_id
rownames(rna_counts) <- rna_genes$gene_id
colnames(rna_counts) <- rownames(rna_clinical)
# Create 5_year_surv column in rna_clinical
five_yr_death <- ifelse(!is.na(rna_clinical$days_to_death), T, F)
days_survive <- ifelse(five_yr_death == "TRUE", 
                       rna_clinical$days_to_death,
                       rna_clinical$days_to_last_follow_up)
# Filter out any remaining NA values for both days f/u and days death
#days_survive <- (days_survive[!is.na(days_survive)])
rna_clinical$five_year_surv <- ifelse(days_survive >= 1825.25, T, F)

# Will be comparing Primary Tumor vs. Normal, so filter out Metastatic Tumor
metastatic_mask <- ifelse(rna_clinical$definition == 'Metastatic', F, T)
rna_clinical <- rna_clinical[metastatic_mask,]
rna_counts <- rna_counts[,metastatic_mask]


```


*2*
Determine and define your experimental design for the differential expression analysis. Control for covariance of age, gender, OR vital_status. Manipulate those columns so that they are ready for differential expression analysis (hint: what kind of variables are they? what data type are they by default? do you need to handle unknown values?)
```{r}
# Will compare this
rna_clinical$definition <- factor(rna_clinical$definition)
# Control for covariance in gender
rna_clinical$gender <- factor(rna_clinical$gender)
# Check for NA values
sum(is.na(rna_clinical$definition)) # 0 NA
sum(is.na(rna_clinical$gender)) # 1 NA
# Remove NA value from any of the above
gender_na_mask <- ifelse(!is.na(rna_clinical$gender), T, F)
rna_clinical <- rna_clinical[gender_na_mask,]
rna_counts <- rna_counts[,gender_na_mask]

# Processing rna_counts df
row_sums <- rowSums(rna_counts)
low_counts_mask <- ifelse(row_sums >= 10, T, F)
# filter out rna_counts + rna_genes for only more than 10
rm_low_counts <- rna_counts[low_counts_mask,]
rm_low_genes <- rna_genes[low_counts_mask,]


```


*3*
Perform the differential expression analysis, create a volcano plot, and interpret what genes from each of the 6 sections means (assuming thresholds of -log10padj = 2 and log2FoldChange = 1)
```{r}
#install.packages('BiocManager')
library(DESeq2)
summary(rna_clinical$definition)
# Running DESeq2
dds <- DESeqDataSetFromMatrix(countData = rm_low_counts,
                             colData = rna_clinical,
                             design = ~gender + definition)
dds_obj <- DESeq(dds)
resultsNames(dds_obj)
results <- results(dds_obj, format = "DataFrame", contrast = c("definition",'Primary solid Tumor', 'Solid Tissue Normal'))

# Analyzing results
results <- data.frame(rm_low_genes$gene_name, results@rownames,
                      results$log2FoldChange, results$pvalue, 
                      results$padj, -log10(results$padj))
colnames(results) <- c('Gene Name', 'Gene ID', 'log2FoldChange',
                       'pvalue', 'padj','-log10(padj)')

# Add gene_id as the row names for results
row.names(results) <- rm_low_genes$gene_id
significant_mask <- ifelse(results$padj < 0.05, T, F)
sig_results <- results[significant_mask,]

# Sort for up regulated results
more_one_mask <- ifelse(sig_results$log2FoldChange > 1, T, F)
positive_results <- sig_results[more_one_mask,]
desc_order <- order(-positive_results$log2FoldChange)
up_reg_results <- positive_results[desc_order,]

# Sort for down regulated results
less_one_mask <- ifelse(sig_results$log2FoldChange < -1, T, F)
negative_results <- sig_results[less_one_mask,]
asc_order <- order(negative_results$log2FoldChange)
down_reg_results <- negative_results[asc_order,]

# Volcano plot
library(EnhancedVolcano)
par(mar=c(1,1,1,1))
EnhancedVolcano(sig_results,
    lab = rownames(sig_results),
    x = 'log2FoldChange',
    y = 'pvalue',
    title = 'Primary Tumor Tissue compared to Normal Tissue',
    pCutoff = 2, # 10e-32
    FCcutoff = 1, # 0.5
    pointSize = 2.0, # 3.0
    labSize = 4.0) # 6.0

```

top-right genes: up-regulated in pt with primary solid tumor, yes significance
bottom-right genes: up-regulated in pt with primary solid tumor, no significance
top-left genes: down-regulated in pt with primary solid tumor, yes significance
bottom-left genes: down-regulated in pt with primary solid tumor, no significance
top-middle genes: not over expressed in either group , yes significance
bottom-middle genes: not over expressed in either group, no significance

Save the picture of the volcano plot (using either ggsave() or right clicking and manually downloading the image and push this .Rmd and the image to GitHub)
